lib/collection/src/collection_manager/holders/proxy_segment.rs

```diff
@@ -454,36 +454,44 @@ impl SegmentEntry for ProxySegment {
     fn rescore_with_formula(
         &self,
         formula_ctx: Arc<FormulaContext>,
-        _wrapped_deleted: Option<&BitSlice>,
         hw_counter: &HardwareCounterCell,
     ) -> OperationResult<Vec<ScoredPoint>> {
-        let deleted_points = self.deleted_points.read();
-
-        // Run rescore in wrapped segment with or without deleted points slice
-        let mut wrapped_results = if deleted_points.is_empty() {
-            self.wrapped_segment.get().read().rescore_with_formula(
-                formula_ctx.clone(),
-                None,
-                hw_counter,
-            )?
-        } else {
-            let deleted_slice = self.deleted_mask.as_deref();
-            self.wrapped_segment.get().read().rescore_with_formula(
-                formula_ctx.clone(),
-                deleted_slice,
-                hw_counter,
-            )?
-        };
+        // Run rescore in wrapped segment
+        let mut wrapped_results = self
+            .wrapped_segment
+            .get()
+            .read()
+            .rescore_with_formula(formula_ctx.clone(), hw_counter)?;
 
         // Run rescore in write segment
-        let mut write_results = self.write_segment.get().read().rescore_with_formula(
-            formula_ctx.clone(),
-            None,
-            hw_counter,
-        )?;
+        let mut write_results = self
+            .write_segment
+            .get()
+            .read()
+            .rescore_with_formula(formula_ctx, hw_counter)?;
 
-        // Just join both results, they will be deduplicated and top-k'd later
-        write_results.append(&mut wrapped_results);
+        // Merge results
+        if self.deleted_points.read().is_empty() {
+            // Just join both results, they will be deduplicated and top-k'd later
+            wrapped_results.append(&mut write_results);
+        } else {
+            // Create a hashmap of write results by point ID for faster lookup
+            // We expect write results to be shorter than wrapped results
+            let mut write_results_map: AHashMap<_, _> = write_results
```

Comment:
I would prefer to avoid allocating HashMaps when possible. This is potentially an expensive operation.

Reply:
Duh, this is much better, thanks